### ISSUE TITLE
Linux version

### ISSUE THREAD

**Issue by hasufell** [KinesisCorporation/SmartSetApps#5]
I bought an advantage 2 for 500€ and can't program it, because there is no release for linux.

**Comment by theAkito** [comment]
It's a mystery to me, that this isn't categorised into a higher priority, since Windows has no FAT16 support, which means you are not able to configure, for example, a Freestyle Edge RGB, because Windows simply can't read the volume, when everything else (including Linux) can.

**Comment by hasufell** [comment]
Kinesis firmware is trash anyway. I hacked my kinesis via https://michael.stapelberg.ch/posts/2020-07-09-kint-kinesis-keyboard-controller/ and now can use proper firmware, namely qmk.

**Comment by jjgalvez** [comment]
Any assistance in getting the edge smartapp to compile in Linux. I compiled all the components, but the project fails.

**Comment by sneznaovca** [comment]
any ways how run soft on linux? @KinesisCorporation

**Comment by zuavra** [comment]
The Kinesis SmartApp runs fine in a Windows 10 virtual machine under VirtualBox. You have to either add the keyboard in the VM's USB configuration, or enable it on the fly while the VM is running from the Devices > USB menu. Make sure the VM has USB emulation activated; USB v1 is sufficient to make it work, no need for 2.x or 3.x explicitly. Make sure you also mount the keyboard's internal flash drive after you start the VM (° + F8).

Doesn't work by default in WINE though.

**Comment by Omniscently** [comment]
> I bought an advantage 2 for 500€ and can't program it, because there is no release for linux.

On-board programming, direct editing of the layout files. You are far from "unable."

**Comment by cartok** [comment]
Who are your customers?.. @KinesisCorporation

**Comment by Fury1** [comment]
Looking to bump this again @KinesisCorporation, a native Linux SmartSetApp would be appreciated. I bought this keyboard because it was marketed as having native Linux support out of the box, the ecosystem should apply to that statement as well.

**Comment by Findarato** [comment]
I personally would love them to just release an optional QMK firmware or optional reading of the configs though a more standard layout so we could just generate our own client to create the text files.

**Comment by Fury1** [comment]
> I personally would love them to just release an optional QMK firmware or optional reading of the configs though a more standard layout so we could just generate our own client to create the text files.

Yea anything would help, I was just looking for a way to visually look at my layout to modify it with a more layered approach.

I am gonna try hacking on the smartsetapp I guess, see if I can get it working without to much effort in my spare time on linux. If it turns into a project, maybe there is a better way to do it utilizing the v-drive text editing method. I'm pretty unfamiliar with all this though...

**Comment by catgoose** [comment]
> > I personally would love them to just release an optional QMK firmware or optional reading of the configs though a more standard layout so we could just generate our own client to create the text files.
> 
> Yea anything would help, I was just looking for a way to visually look at my layout to modify it with a more layered approach.
> 
> I am gonna try hacking on the smartsetapp I guess, see if I can get it working without to much effort in my spare time on linux. If it turns into a project, maybe there is a better way to do it utilizing the v-drive text editing method. I'm pretty unfamiliar with all this though...

It seems like you can edit config files without the software:

https://www.reddit.com/r/MechanicalKeyboards/comments/p1mtf8/comment/hce0d5l/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by Fury1** [comment]
> > > I personally would love them to just release an optional QMK firmware or optional reading of the configs though a more standard layout so we could just generate our own client to create the text files.
> > 
> > 
> > Yea anything would help, I was just looking for a way to visually look at my layout to modify it with a more layered approach.
> > I am gonna try hacking on the smartsetapp I guess, see if I can get it working without to much effort in my spare time on linux. If it turns into a project, maybe there is a better way to do it utilizing the v-drive text editing method. I'm pretty unfamiliar with all this though...
> 
> It seems like you can edit config files without the software:
> 
> https://www.reddit.com/r/MechanicalKeyboards/comments/p1mtf8/comment/hce0d5l/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Right, this is known. However, again it would be nice to have the visual aid of the smartset software when performing remaps. I myself have been using the onboard remapping features but sometimes find it hard to visualize my current layered customizations. I do like the keyboard on a personal level, but I purchased it on the premise of having the advertised linux support which turns out to only be partially true in regard to the smartset ecosystem on the wired version.

**Comment by catgoose** [comment]
> > > > I personally would love them to just release an optional QMK firmware or optional reading of the configs though a more standard layout so we could just generate our own client to create the text files.
> > > 
> > > 
> > > Yea anything would help, I was just looking for a way to visually look at my layout to modify it with a more layered approach.
> > > I am gonna try hacking on the smartsetapp I guess, see if I can get it working without to much effort in my spare time on linux. If it turns into a project, maybe there is a better way to do it utilizing the v-drive text editing method. I'm pretty unfamiliar with all this though...
> > 
> > 
> > It seems like you can edit config files without the software:
> > https://www.reddit.com/r/MechanicalKeyboards/comments/p1mtf8/comment/hce0d5l/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
> 
> Right, this is known. However, again it would be nice to have the visual aid of the smartset software when performing remaps. I myself have been using the onboard remapping features but sometimes find it hard to visualize my current layered customizations. I do like the keyboard on a personal level, but I purchased it on the premise of having the advertised linux support which turns out to only be partially true in regard to the smartset ecosystem on the wired version.

Yeah I understand.  I am still waiting on my keyboard in the mail so in the meantime I'm going to play around with Lazarus and see if I can figure anything out.

**Comment by JackMBurch** [comment]
Yeah, this is kinda surprising as I'm sure a large part of their customer base is linux based.

**Comment by Fury1** [comment]
I have been working on this a bit. To preface;  _if it sounds like I don't know what I'm doing, its because I don't know what I'm doing and I have never touched pascal or Lazarus in my life..._

With that said, I have Lazarus setup on Linux and compiled with all third party packages required, there were a few modifications needed in the source code for Lazarus v3. It looks like the SmartSet project was built using Lazarus v2. Why am I using 3 then? Because the README said to install the latest version. The more I look at things it _feels_ like this repo is very out of date.

~~I also am not sure how to compile the actual program....the **SmartSetAdv360** folder is not a complete Lazarus project as there is no `.lpi` file, there is also nothing listed about it under [Installation](https://github.com/KinesisCorporation/SmartSetApps#source-folders). I tried looking at some of the commits for clues but there are 360 commit messages kind of spread around.~~

I wanna say that the **SmartSetMasterOffice** folder is the app to compile. 

~~I'm a bit stuck at the moment because I don't really know anything about pascal and the Lazarus IDE seems like its an IDE and the environment itself for the project.~~

From what I can tell there is no Linux compile target setup in the program. 

MacOS is targeted which is close to Linux, however, it seems like depending on the OS the pascal units will be switched out at compile time like a preprocessor directive in C++. The problem now is the current MacOS HTTP client is `Components/ns_url_request.pas` which uses a `$modeswitch ObjectiveC1` which is not compatible with my `Linux` compile target per this message:

`ns_url_request.pas(16,2) Error: The feature "Objective-C" is not, or not yet, supported on the selected target platform`

I assume this means a compatible Linux http client is needed and will need to be linked to.

Anybody want to chime in? @KinesisCorporation

**Comment by KinesisCorporation** [comment]
@Fury1  The Advantage2 and Freestyle Pro versions of the SmartSet App are in the process of being deprecated. We will be migrating Advantage2 support to the platform running the Advantage360 SmartSet.

Unfortunately Kinesis is not in a position to further support Linux development of any version of the SmartSet App. Linux users are encouraged to simply edit the text files directly. The tokens and syntax are described in detail in the Adv2 User Manual linked below. For sophisticated users its significantly quicker, easier and more powerful to do it that way in our opinion.

https://kinesis-ergo.com/support/advantage2/#manuals

**Comment by hasufell** [comment]
I'm glad support managed to reply after... just 4 years!

**Comment by michael-hart-github** [comment]
> Unfortunately Kinesis is not in a position to further support Linux development of any version of the SmartSet App. Linux users are encouraged to simply edit the text files directly. The tokens and syntax are described in detail in the Adv2 User Manual linked below. For sophisticated users its significantly quicker, easier and more powerful to do it that way in our opinion.

If that is the case, please update the material on your site to reflect this fact. As is, it merely looks like you want Linux users to download and compile from source, _not_ that you don't even support it.

![Image](https://github.com/user-attachments/assets/5a69f3e1-bd0b-4567-a7df-a697af2b9133)

https://kinesis-ergo.com/support/kb360/

Regardless, thank you for the suggestion you have provided. Hopefully it will work as a substitution.

**Comment by KinesisCorporation** [comment]
Hi Michael,

We've always tried to be very clear that the SmartSet App itself is Windows/Mac, but that Onboard Programming and Direct Editing are fully supported on all operating systems. But obviously we fell short. It was only by customer request that we even made the GitHub available for Linux users. Our hope was that would be better than nothing but obviously that was not the case.

We will work on syncing up our website to match the current guidance. 

And please send us a ticket if you have any problems with Direct Editing.

**Comment by aMytho** [comment]
I have an advantage 360 and I was able to get the smartset app running with wine 10 stable on kubuntu 24.10. There are some graphical issues, (random lines being drawn, may be due to my driver issues), but it was functional. 

I had to mount the keyboard v drive before opening the app. I remapped 1 key and saved it. There was an error message about storage permissions being denied, but I closed the warning box and it saved the changes anyway.

Its definitely not ideal, but for those who really need a GUI you could probably make it work. I  **strongly** suggest making a backup of your keyboard files before trying this. I also haven't tried anything other than remapping a key, but since the app just generates text files I would assume it works fine. I was able to view my layers so at the very least it makes it easier to visualize with the onboard programming.
